packages/tasit-action/src/TasitAction.test.js

```diff
@@ -84,109 +77,292 @@ describe("Contract", () => {
-    simpleStorage = simpleStorage.setWallet(wallet);
+    beforeEach("assign a wallet to the contract", () => {
+      expect(() => {
+        simpleStorage.setWallet(wallet);
+      }).to.not.throw();
+    });
 
-    var rand = Math.floor(Math.random() * Math.floor(1000)).toString();
-    const subscription = simpleStorage.setValue(rand);
+    afterEach("waiting for message/tx confirmation", async () => {
+      if (subscription) {
+        await subscription.waitForMessage();
+        subscription.removeAllListeners();
+      }
+    });
 
-    const onMessage = async message => {
-      // message.data = Contents of the message.
-      const { data } = message;
-      const { confirmations } = data;
+    it("should throw when subscribing with invalid trigger", async () => {
+      subscription = simpleStorage.setValue("hello world");
+      return expect(subscription.on("invalid", () => {})).to.be.rejected;
+    });
 
-      if (confirmations === 7) {
-        subscription.removeAllListeners();
+    it("should throw when subscribing without callback", async () => {
+      subscription = simpleStorage.setValue("hello world");
+      return expect(subscription.on("confirmation")).to.be.rejected;
+    });
+
+    it("should call a write contract method (send tx) - imediate subscription", async () => {
+      var rand = Math.floor(Math.random() * Math.floor(1000)).toString();
+      subscription = simpleStorage.setValue(rand);
+      const fakeFn = sinon.fake();
+
+      const onMessage = async message => {
+        const { data } = message;
+        const { confirmations } = data;
+
+        if (confirmations >= 7) {
+          fakeFn();
+
+          subscription.removeAllListeners();
+
+          const value = await simpleStorage.getValue();
+          expect(value).to.equal(rand);
+
+          // FIXME: UnhandledPromiseRejectionWarning
+          //expect(1).to.equal(2);
+        }
+      };
+
+      await subscription.on("confirmation", onMessage);
+
+      await mineBlocks(simpleStorage.getProvider(), 15);
+
+      expect(fakeFn.called).to.be.true;
+    });
+
+    it("should call a write contract method (send tx) - late subscription", async () => {
+      var rand = Math.floor(Math.random() * Math.floor(1000)).toString();
+      subscription = simpleStorage.setValue(rand);
+      const fakeFn = sinon.fake();
+
+      await mineBlocks(simpleStorage.getProvider(), 15);
+
+      const onMessage = async message => {
+        const { data } = message;
+        const { confirmations } = data;
+
+        if (confirmations >= 7) {
+          fakeFn();
+
+          subscription.removeAllListeners();
+
+          const value = await simpleStorage.getValue();
+          expect(value).to.equal(rand);
+        }
+      };
+
+      await subscription.on("confirmation", onMessage);
+
+      await mineBlocks(simpleStorage.getProvider(), 20);
+
+      expect(fakeFn.called).to.be.true;
+    });
+
+    it("should remove listener after timeout", async () => {
```

Comment:
I think the name of the test should include the fact that the error function is called. Because I think that's what developers using the SDK will make use of - they'll have the UI change state after this timeout event fires

Reply:
After PR https://github.com/tasitlabs/TasitSDK/pull/102, the listener is not more deleted. The test case description is now `should call error listener after timeout`

Reply:
Perfect. I'll resolve this then.